Content locking

distantnative opened the pull request:
# Todo
- [x] Implement for site and users
- [x] Figure out a concept for files and implement
- [x] Protect models from being deleted/renamed etc. while locked
- [x] Fix cmd+S
- [x] Disable form buttons during saving (on slow connections)
- [x] Code cleaning
- [x] Add unit tests for backend part
- [x] Better styling (I'll leave that for @bastianallgeier)
- [x] Translation strings
- [ ] Changelog: Add warning that locking only works across different users for now, not for different clients of the same user

# Related issue
#1634

lukasbestle commented on config/api/routes/lock.php:
```diff
@@ -0,0 +1,50 @@
+        'pattern' => '(:all)/lock',
+        'method'  => 'PATCH',
+        'action'  => function (string $path) {
+            return $this->parent($path)->lock($this->requestBody('user'), $this->language());
```
I think it would be better to use the currently authenticated user here (as well as for the `unlock` routes below), not a user that gets passed from the client. After all it should always only be possible to lock an object to the current user.

lukasbestle commented on config/api/routes/lock.php:
```diff
@@ -0,0 +1,50 @@
+        'pattern' => '(:all)/lock',
+        'method'  => 'DELETE',
+        'action'  => function (string $path) {
+            return $this->parent($path)->lock(null, $this->language());
```
Shouldn't this route check whether the lock user and the current authenticated user are the same? Removing the lock without triggering `unlock` should only be possible for the user who locked the object, no?

Also see my comment for the `src/Cms/HasLock.php` trait.

lukasbestle commented on src/Cms/HasLock.php:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        // write: set
+        return $this->update([
```
I think this should delete the `lock` field at the same time. The client would then only need one request and not one to remove the lock and then one to call `unlock` afterwards.

lukasbestle commented on panel/src/components/Forms/FormButtons.vue:
```diff
@@ -22,29 +58,174 @@
+  },
   computed: {
+    canUnlock() {
+      return this.lock.time + (60 * 2) < this.now.unix();
```
Shouldn't this check be handled by the backend? Otherwise it's possible to unlock the page with a manual API request. Instead, the backend could pass this boolean `canUnlock` value back with the lock request.

Another advantage of handling this in the backend: The interval could be made configurable at a later point if needed.

distantnative replied:
👍

lukasbestle commented on panel/src/components/Forms/FormButtons.vue:
```diff
@@ -22,29 +58,174 @@
+      // if user started to make changes,
+      // start setting lock on each heartbeat
+      if (current === true && previous == false) {
+        this.onHeartbeat(this.setLock, 30);
```
Why is the lock set over and over?

distantnative replied:
Since we base it on the time stamp, we have to keep updating it as long as the user keeps editing.

lukasbestle replied:
Ah, of course! :)

lukasbestle commented on panel/src/components/Forms/FormButtons.vue:
```diff
@@ -22,29 +58,174 @@
+
+        // if a lock is set and it's not from the current user
+        // dispatch lock information to store
+        if (response.lock === true && response.user !== this.user) {
```
What happens if the user locks an object and then tries to access the same object with the same account, but from a different device? Won't it appear unlocked there even though the local storage is only on the other device?

distantnative replied:
Yes, but honestly I feel like preventing this might become veeery complex. So I left that agency with the user.

lukasbestle replied:
What we could do: If the lock is set to the current user but there's nothing in the local storage on the current device, treat the lock like a lock of a different user. The message would be something like "this page/file is locked by you on a different device".

distantnative replied:
@lukasbestle with the current implementation we can't do it at the moment, since the backend returns no lock for the current user if the lock is by himself/herself.

lukasbestle commented on panel/src/components/Forms/FormButtons.vue:
```diff
@@ -22,29 +58,174 @@
+          this.$store.dispatch("form/lock", {
+            user: response.user,
+            email: response.email,
+            time: parseInt(response.time)
```
This should use the second argument of `parseInt(response.time, 10)` to avoid browser compat issues.

distantnative replied:
👍

lukasbestle commented on panel/src/components/Forms/FormButtons.vue:
```diff
@@ -22,29 +58,174 @@
+      Object.keys(this.unlock).forEach(key => {
+        content += key + ": \n" + this.unlock[key];
+        content += "\n\n----\n\n";
+      });
```
The backend implementation in `Data\Txt` has a few special cases that are not covered here. But I don't think we should duplicate the logic in the frontend either. Two options:

- Ignore the special cases, which means that the downloaded file might not be usable 1:1, but is a well-enough backup.
- Create a helper API route that can be passed the `this.unlock` object and that will then throw the object at `Data\Txt::encode()` and return the resulting string. Advantage: That helper route could actually already trigger the download to make the link hack obsolete.

distantnative replied:
The download file isn’t usable 1:1 anyways at the moment since it only includes the changes.

lukasbestle replied:
That's right. Maybe we can keep the basic implementation for now and extend it later if there's demand for it.

lukasbestle commented on src/Cms/File.php:
```diff
@@ -228,6 +228,28 @@ public function contentFileName(): string
+     * @internal
+     * @return string
+     */
+    public function contentLockFile(): string
```
Are those two methods left-overs from the old implementation?

distantnative replied:
Indeed.

lukasbestle commented on src/Cms/HasLock.php:
```diff
@@ -0,0 +1,73 @@
+trait HasLock
+{
+
+    public function lock($user = false, string $language = null)
```
In my opinion, the `lock()` method and *especially* the `unlock()` method do way too much stuff. Also it's completely intransparent which combination of arguments gets or sets stuff. If called incorrectly (e.g. `null` instead of `false`), the methods could break in unexpected ways.

I see you had different get and set methods previously. To be honest I'd return to separate `lock()` and `getLock()` methods (same for unlocking). The setters could accept `string? $user`, which would also accept `null` for "unset".

distantnative replied:
As you saw, I had those and then switched to having only one for each as suggested by @bastianallgeier on Slack. I preferred explicit ones as well, so let’s hear from Bastian.

lukasbestle commented on src/Cms/HasLock.php:
```diff
@@ -0,0 +1,73 @@
+        // write: set
+        return $this->update([
+            'lock' => [
+                'user' => $user,
```
There's no type validation on this. Fixes itself if the method only accepts `string?` (see above).

lukasbestle commented on src/Cms/HasLock.php:
```diff
@@ -0,0 +1,73 @@
+
+        // read
+        if ($user === null) {
+            return $data;
```
I think the getter should also accept a `$user` param that (if set) returns whether the passed user is in the unlock data (`in_array($user, $data)`).

lukasbestle commented on src/Cms/HasLock.php:
```diff
@@ -0,0 +1,73 @@
+
+        // write: set
+        return $this->update([
+            'unlock' => array_merge($data, [$user])
```
I assume you wanted to fit this into one line, but I'd say it's more readable to use `$data[] = $user` above instead.

bastianallgeier reviewed (commented):
Looks great so far! I only wonder if it makes sense to run the heartbeat all the time, no matter if it's needed or not. I can't really tell what this means for performance.

lukasbestle commented on panel/src/components/Sections/FilesSection.vue:
```diff
@@ -86,34 +86,45 @@ export default {
+      this.$api.get(url).then(response => {
+
+        // restrict actions if file is locked
+        if (response.locke && ["download", "edit"].includes(action) === false) {
```
Shouldn't this be `response.locked`?

lukasbestle commented on src/Cms/ModelLock.php:
```diff
@@ -0,0 +1,319 @@
+    /**
+     * @param ModelWithContent $model
+     */
+    public function __construct($model)
```
This could use a type hint:

```php
public function __construct(ModelWithContent $model)
```

lukasbestle commented on src/Cms/ModelLock.php:
```diff
@@ -0,0 +1,319 @@
+    public function __construct($model)
+    {
+        $this->model = $model;
+        $this->data  = $this->read();
```
Reading the lock file for each model will lead to issues if the lock of multiple models in the same directory (e.g. the page and a file or multiple files) is modified in the same request as each model will load the lock file once and then write it back, overwriting the lock changes from other models.

I think the most robust implementation would be a global `$kirby->locks()` collection that acts as a proxy:

- `$kirby->locks()->get(ModelWithContent $model)` would return the data array for a model. So all the model ID and lock file reading stuff would get handled by this (determine the correct lock file based on the model, open it, cache the contents of that lock file in an instance property, extract the array of the model by ID, return it).
- `$kirby->locks()->set(ModelWithContent $model, array $data)` would update the cache in the instance property and then write the lock file again.
- The `ModelLock` class could get slimmed down a lot because it wouldn't need to handle all the low-level file operations and model ID stuff. In the constructor, it would just `get()` its data with `$model->kirby()->locks()->get($model)` and on each change it would call `$this->kirby()->locks()->set($this->model(), $this->data)` (where `$this->data` no longer contains all lock file contents, but one level fewer: just the data of the current model).

I hope my explanation was good enough. If not, please let me know. :)

distantnative replied:
Wouldn’t this increase the likelihood of race conditions as they’re now affected by different users editing any model at the same time while currently it’s only while editing models from the same content folder?

lukasbestle replied:
There would still be a lock file per folder, so no difference there. Those different lock files would just be managed by a global class.

Keeping this global would also make it a lot easier to add FS locking later.

distantnative replied:
Then I don’t understand the advantage yet. The Panel send individual requests per lock. So whether this is handled in the App class or in the Model class, it would only handle the data of one model since each API requests initialized Kirby for itself.

lukasbestle replied:
For API requests yes, but if locks are modified manually in a script it will break.

distantnative replied:
✅

lukasbestle commented on src/Cms/ModelLock.php:
```diff
@@ -0,0 +1,319 @@
+     */
+    protected function id(): string
+    {
+        return '/' . $this->model()->id();
```
What was the reason for the leading slash?

distantnative replied:
Makes it work for Site

lukasbestle replied:
Ah, of course.

lukasbestle reviewed (dismissed):
## Race condition bug

**How to reproduce:**

1. Make changes in browser A.
2. Quickly make changes in browser B (different user) before browser B knows of the lock.

**Result:**

Browser B can no longer access the corrupt model (I get kicked to the Panel home page every time I try to visit the corrupt page). There is no way to recover from this in browser B (except clearing local storage manually).

**Reason for this bug:**

The state of browser B will be corrupt: There will be local changes although a lock of another user exists, so browser B will try to create a lock but will fail because a lock exists (API exception: "... is already locked").

**Possible solution:**

Handle this corrupt state in the frontend exactly like the "unlock" state: Browser B would now present the "Your unsaved changes have been unlocked by another user." message and offer to download the local changes. If confirmed, the local changes would get cleared and the local state would go to "The file is currently being edited by \<Browser A\>".

Commit:
Implement content locking

# Conflicts:
#	panel/src/components/Sections/PagesSection.vue

lukasbestle committed:
Cache ContentLocks instance as singleton

Otherwise there would be one instance for each model, which wouldn’t solve the issue of multiple locks overwriting each other in the same files.

lukasbestle committed:
Smaller backend code refactoring & doc blocks

Commit:
Fix values/changes state in form store

lukasbestle committed:
File system locking for .lock files

lukasbestle commented on panel/src/components/Forms/FormButtons.vue:
```diff
@@ -27,24 +62,140 @@ export default {
+    },
+    setLock() {
+      this.$api.patch(this.$route.path + "/lock", null, null, true).catch(() => {
+        this.$store.dispatch("form/revert", this.id);
```
This seems to trigger a request to the `DELETE /lock` API endpoint, which results in another Exception ("The content lock can only be removed by the user who created it.") in the browser console. Does not break the behavior, only a cosmetic issue. If you can fix it without too much effort, that would be great. Otherwise leave it like this. :)

distantnative replied:
I don't think there is an easy way out of this... It's triggered here: https://github.com/getkirby/kirby/pull/1718/files#diff-53f37eaee7ac16a1f56969f2cd5b1a56R95
We need that part, so that a lock gets removed if the user makes a change (setting the lock) and then manually changes the value back to the original. The Panel recognises this as no changes anymore and so the lock should also be removed. But it is also triggered here when reverting the changes directly when discovered that the page is already locked.

Option would be to remove that revert line but then user B would see his/her few changes still in the disabled form inputs (disabled because of the lock by user A), even though they will be ignored immediately when user A saves or reverts.

lukasbestle replied:
That's what I suspected. The line definitely has its right to exist, I thought of a check like "only try to remove the lock if there was one before". But if that's not doable, feel free to resolve this conversation. Doesn't matter that much.

lukasbestle reviewed (commented):
Awesome! I have reviewed your changes and also made some manual tests. Seems to work pretty great now.

I have made some improvements to the backend code and also implemented file system locking.

Only two things left:

- The cosmetic bug in the frontend
- I think we could use a "relative" model ID so that the page/site model gets locked under `/` and the files under `/bla.jpg`. So basically the IDs would be relative to the model's content directory. This allows moving the directory in the file system and renaming pages on levels above. Currently all those locks would break as the model ID would no longer match.

Besides that: Awesome work! After all the open todos (testing etc.) are done, Bastian can do his final review. From my side everything is 👍.

Commit:
Small code style fixes

Commit:
Add tests for content locking

bastianallgeier committed:
Fix invalid invisible characters

bastianallgeier committed:
Fix design of the wrapping container

bastianallgeier committed:
More translations

bastianallgeier committed:
Improve order of bars in form buttons

bastianallgeier committed:
Don’t inherit html attributes for sections
